man/g_enemat.1
.TH g_enemat 1 "Tue 15 May 2001"
.SH NAME
g_enemat
.B VERSION 3.0
.SH SYNOPSIS
\f3g_enemat\fP
.BI "-f" " ener.edr "
.BI "-groups" " groups.dat "
.BI "-eref" " eref.dat "
.BI "-emat" " emat.xpm "
.BI "-etot" " energy.xvg "
.BI "-[no]h" ""
.BI "-[no]X" ""
.BI "-nice" " int "
.BI "-b" " time "
.BI "-e" " time "
.BI "-dt" " time "
.BI "-[no]w" ""
.BI "-[no]sum" ""
.BI "-skip" " int "
.BI "-[no]mean" ""
.BI "-nlevels" " int "
.BI "-max" " real "
.BI "-min" " real "
.BI "-[no]coul" ""
.BI "-[no]coulr" ""
.BI "-[no]coul14" ""
.BI "-[no]lj" ""
.BI "-[no]lj14" ""
.BI "-[no]bham" ""
.BI "-[no]free" ""
.BI "-temp" " real "
.SH DESCRIPTION
g_enemat extracts an energy matrix from an energy file.
With 
.B -groups
a file must be supplied with on each
line a group to be used. For these groups a matrices of
interaction energies will be calculated. Also the total
interaction energy energy per group is calculated.


An approximation of the free energy is calculated using:
E(free) = E0 + kT log( exp((E-E0)/kT) ), where ''
stands for time-average. A file with reference free energies
can be supplied to calculate the free energy difference
with some reference state. Group names (e.g. residue names
in the reference file should correspond to the group names
as used in the 
.B -groups
file, but a appended number
(e.g. residue number)in the 
.B -groups
will be ignored
in the comparison.
.SH FILES
.BI "-f" " ener.edr" 
.B Input, Opt.
 Generic energy: edr ene 

.BI "-groups" " groups.dat" 
.B Input
 Generic data file 

.BI "-eref" " eref.dat" 
.B Input, Opt.
 Generic data file 

.BI "-emat" " emat.xpm" 
.B Output
 X PixMap compatible matrix file 

.BI "-etot" " energy.xvg" 
.B Output
 xvgr/xmgr file 

.SH OTHER OPTIONS
.BI "-[no]h"  "    no"
 Print help info and quit

.BI "-[no]X"  "    no"
 Use dialog box GUI to edit command line options

.BI "-nice"  " int" " 19" 
 Set the nicelevel

.BI "-b"  " time" "     -1" 
 First frame (ps) to read from trajectory

.BI "-e"  " time" "     -1" 
 Last frame (ps) to read from trajectory

.BI "-dt"  " time" "     -1" 
 Only use frame when t MOD dt = first time (ps)

.BI "-[no]w"  "    no"
 View output xvg, xpm, eps and pdb files

.BI "-[no]sum"  "    no"
 Sum the energy terms selected rather than display them all

.BI "-skip"  " int" " 0" 
 Skip number of frames between data points

.BI "-[no]mean"  "   yes"
 with -groups calculates matrix of mean energies in stead of matrix for each timestep

.BI "-nlevels"  " int" " 20" 
 number of levels for matrix colors

.BI "-max"  " real" "  1e+20" 
 max value for energies

.BI "-min"  " real" " -1e+20" 
 min value for energies

.BI "-[no]coul"  "   yes"
 calculate Coulomb SR energies

.BI "-[no]coulr"  "    no"
 calculate Coulomb LR energies

.BI "-[no]coul14"  "    no"
 calculate Coulomb 1-4 energies

.BI "-[no]lj"  "   yes"
 calculate Lennard-Jones SR energies

.BI "-[no]lj14"  "    no"
 calculate Lennard-Jones 1-4 energies

.BI "-[no]bham"  "    no"
 calculate Buckingham energies

.BI "-[no]free"  "   yes"
 calculate free energy

.BI "-temp"  " real" "    300" 
 reference temperature for free energy calculation



src/tools/g_enemat.c
/*
 *       $Id$
 *
 *       This source code is part of
 *
 *        G   R   O   M   A   C   S
 *
 * GROningen MAchine for Chemical Simulations
 *
 *            VERSION 1.6
 * 
 * Copyright (c) 1991-1997
 * BIOSON Research Institute, Dept. of Biophysical Chemistry
 * University of Groningen, The Netherlands
 * 
 * Please refer to:
 * GROMACS: A message-passing parallel molecular dynamics implementation
 * H.J.C. Berendsen, D. van der Spoel and R. van Drunen
 * Comp. Phys. Comm. 91, 43-56 (1995)
 *
 * Also check out our WWW page:
 * http://rugmd0.chem.rug.nl/~gmx
 * or e-mail to:
 * gromacs@chem.rug.nl
 *
 * And Hey:
 * Good ROcking Metal Altar for Chronical Sinners
 */
static char *SRCID_g_enemat_c = "$Id$";

#include <string.h>
#include <math.h>
#include "typedefs.h"
#include "fatal.h"
#include "vec.h"
#include "smalloc.h"
#include "enerio.h"
#include "statutil.h"
#include "assert.h"
#include "names.h"
#include "copyrite.h"
#include "macros.h"
#include "xvgr.h"
#include "gstat.h"
#include "physics.h"
#include "matio.h"

static int search_str2(int nstr,char **str,char *key)
{
  int  i,n;

  /* Linear search */
  n=0;
  while( (n<strlen(key)) && ((key[n]<'0') || (key[n]>'9')) )
    n++;
  for(i=0; (i<nstr); i++) 
    if (strncasecmp(str[i],key,n)==0)
      return i;

  return -1;
}

int main(int argc,char *argv[])
{
  static char *desc[] = {
    "g_enemat extracts an energy matrix from an energy file.",
    "With [BB]-groups[bb] a file must be supplied with on each",
    "line a group to be used. For these groups a matrices of",
    "interaction energies will be calculated. Also the total",
    "interaction energy energy per group is calculated.[PAR]",
    "An approximation of the free energy is calculated using:",
    "E(free) = E0 + kT log( <exp((E-E0)/kT)> ), where '<>'",
    "stands for time-average. A file with reference free energies",
    "can be supplied to calculate the free energy difference",
    "with some reference state. Group names (e.g. residue names",
    "in the reference file should correspond to the group names",
    "as used in the [BB]-groups[bb] file, but a appended number",
    "(e.g. residue number)in the [BB]-groups[bb] will be ignored",
    "in the comparison."
  };
  static bool bSum=FALSE;
  static bool bMeanEmtx=TRUE;
  static int  skip=0,nlevels=20;
  static real cutmax=1e20,cutmin=-1e20,reftemp=300.0;
  static bool bCoul=TRUE,bCoulLR=FALSE,bCoul14=FALSE;
  static bool bLJ=TRUE,bLJ14=FALSE,bBham=FALSE,bFree=TRUE;
  t_pargs pa[] = {
    { "-sum",  FALSE, etBOOL, &bSum,
      "Sum the energy terms selected rather than display them all" },
    { "-skip", FALSE, etINT,  &skip,
      "Skip number of frames between data points" },
    { "-mean", FALSE, etBOOL, &bMeanEmtx,
      "with -groups calculates matrix of mean energies in stead of "
      "matrix for each timestep" },
    { "-nlevels", FALSE, etINT, &nlevels,"number of levels for matrix colors"},
    { "-max",FALSE, etREAL, &cutmax,"max value for energies"},
    { "-min",FALSE, etREAL, &cutmin,"min value for energies"},
    { "-coul", FALSE, etBOOL, &bCoul,"calculate Coulomb SR energies"},
    { "-coulr", FALSE, etBOOL, &bCoulLR,"calculate Coulomb LR energies"},
    { "-coul14",FALSE, etBOOL, &bCoul14,"calculate Coulomb 1-4 energies"},
    { "-lj", FALSE, etBOOL, &bLJ,"calculate Lennard-Jones SR energies"},
    { "-lj14",FALSE, etBOOL, &bLJ14,"calculate Lennard-Jones 1-4 energies"},
    { "-bham",FALSE, etBOOL, &bBham,"calculate Buckingham energies"},
    { "-free",FALSE, etBOOL, &bFree,"calculate free energy"},
    { "-temp",FALSE, etREAL, &reftemp,
      "reference temperature for free energy calculation"}
  };
  /* We will define egSP more energy-groups: 
     egTotal (total energy) */
#define egTotal egNR
#define egSP 1
  bool       egrp_use[egNR+egSP];
  FILE       *in,*out;
  int        timecheck;
  XDR        xdr;
  t_energy   *ee;
  t_drblock  dr;
  int        teller=0,nre,step;
  real       t,sum;
  bool       bCont,bEDR,bRef;
  bool       bCutmax,bCutmin;
  real       **eneset,*time=NULL;
  int        *set,i,j,k,prevk,m,n,nset,nenergy;
  char       **enm,**groups;
  char       groupname[255],fn[255];
  int        ngroups;
  t_rgb      rlo,rhi,rmid;
  real       emax,emid,emin;
  real       ***emat,**etot,*groupnr;
  double     beta,expE,**e,*eaver,*efree;
  char       label[234];
  char       **ereflines,**erefres;
  real       *eref,*edif;
  int        neref;
  
  t_filenm   fnm[] = {
    { efENE, "-f", NULL, ffOPTRD },
    { efEDR, "-d", NULL, ffOPTRD },
    { efDAT, "-groups", "groups.dat", ffREAD },
    { efDAT, "-eref",   "eref.dat", ffOPTRD },
    { efXPM, "-emat",   "emat", ffWRITE },
    { efXVG, "-etot",   "energy", ffWRITE }
  };
#define NFILE asize(fnm)

  CopyRight(stderr,argv[0]);
  parse_common_args(&argc,argv,PCA_CAN_VIEW | PCA_CAN_TIME,TRUE,
		    NFILE,fnm,asize(pa),pa,asize(desc),desc,0,NULL);
  
  egrp_use[egCOUL]=bCoul;
  egrp_use[egLJ]=bLJ;
  egrp_use[egBHAM]=bBham;
  egrp_use[egLR]=bCoulLR;
  egrp_use[egCOUL14]=bCoul14;
  egrp_use[egLJ14]=bLJ14;
  egrp_use[egTotal]=TRUE;
  
  bRef=opt2bSet("-eref",NFILE,fnm);
  bEDR=ftp2bSet(efEDR,NFILE,fnm);
  if (bEDR) {
    xdropen(&xdr,ftp2fn(efEDR,NFILE,fnm),"r");
    enm=NULL;
    edr_nms(&xdr,&nre,&enm);
  }
  else {
    in=ftp2FILE(efENE,NFILE,fnm,"r");
    rd_ener_nms(in,&nre,&enm);
  }
  
  if (nre == 0) {
    fprintf(stderr,"No energies!\n");
    exit(1);
  }
  
  bCutmax=opt2parg_bSet("-max",asize(pa),pa);
  bCutmin=opt2parg_bSet("-min",asize(pa),pa);

  snew(ee,nre);
  nenergy = 0;

  /* Read groupnames from input file and construct selection of 
     energy groups from it*/
  
  fprintf(stderr,"Will read groupnames from inputfile\n");
  ngroups = get_lines(opt2fn("-groups",NFILE,fnm), &groups);
  fprintf(stderr,"Read %d groups\n",ngroups);
  snew(set,sqr(ngroups)*egNR/2);
  n=0;
  prevk=0;
  for (i=0; (i<ngroups); i++) {
    fprintf(stderr,"\rgroup %d",i);
    for (j=i; (j<ngroups); j++)
      for(m=0; (m<egNR); m++) 
	if (egrp_use[m]) {
	  sprintf(groupname,"%s:%s-%s",egrp_nm[m],groups[i],groups[j]);
#ifdef DEBUG
	  fprintf(stderr,"\r%-15s %5d",groupname,n);
#endif
	  for(k=prevk; (k<prevk+nre); k++)
	    if (strcmp(enm[k%nre],groupname) == 0) {
	      set[n++] = k;
	      break;
	    }
	  if (k==prevk+nre)
	    fprintf(stderr,"WARNING! could not find group %s (%d,%d)"
		    "in energy file\n",groupname,i,j);
	  else
	    prevk = k;
	}
  }
  fprintf(stderr,"\n");
  nset=n;
  snew(eneset,nset+1);
  fprintf(stderr,"Will select half-matrix of energies with %d elements\n",n);

  /* Start reading energy frames */  
  step     = 0;
  t        = 0;
  do {
    do {
      if (bEDR)
	bCont=edr_io(&xdr,&t,&step,&nre,ee,&dr);
      else
	bCont=rd_ener(in,&t,&step,ee,&dr);
      if (bCont) {
	timecheck=check_times(t);
	
	/* It is necessary for statistics to start counting from 1 */
	step += 1; 
      }
      
    } while (bCont && (timecheck < 0));
    
    if (timecheck == 0) {
#define DONTSKIP(cnt) (skip) ? ((cnt % skip) == 0) : TRUE
      
      if (bCont) {
	fprintf(stderr,"\rRead frame: %d, Time: %.3f",teller,t);
	
	if ((nenergy % 1000) == 0) {
	  srenew(time,nenergy+1000);
	  for(i=0; (i<=nset); i++)
	    srenew(eneset[i],nenergy+1000);
	}
	time[nenergy] = t;
	sum=0;
	for(i=0; (i<nset); i++) {
	  eneset[i][nenergy] = ee[set[i]].e;
	  sum+=ee[set[i]].e;
	}
	if (bSum) 
	  eneset[nset][nenergy] = sum;
	nenergy++;
      }
      teller++;
    }
  } while (bCont && (timecheck == 0));
  
  fprintf(stderr,"\n");

  fprintf(stderr,"Will build energy half-matrix of %d groups, %d elements, "
	  "over %d frames\n",ngroups,nset,nenergy);
  
  snew(emat,egNR+egSP);
  for(j=0; (j<egNR+egSP); j++)
    if (egrp_use[m]) {
      snew(emat[j],ngroups);
      for (i=0; (i<ngroups); i++)
	snew(emat[j][i],ngroups);
    }
  snew(groupnr,ngroups);
  for (i=0; (i<ngroups); i++)
    groupnr[i] = i+1;
  rlo.r  = 1.0, rlo.g  = 0.0, rlo.b  = 0.0;
  rmid.r = 1.0, rmid.g = 1.0, rmid.b = 1.0;
  rhi.r  = 0.0, rhi.g  = 0.0, rhi.b  = 1.0;
  if (bMeanEmtx) {
    snew(e,ngroups);
    for (i=0; (i<ngroups); i++) 
      snew(e[i],nenergy);
    n = 0;
    for (i=0; (i<ngroups); i++) {
      for (j=i; (j<ngroups); j++) {
	for (m=0; (m<egNR); m++) {
	  if (egrp_use[m]) {
	    for (k=0; (k<nenergy); k++) {
	      emat[m][i][j] += eneset[n][k];
	      e[i][k] += eneset[n][k];/* *0.5; */
	      e[j][k] += eneset[n][k];/* *0.5; */
	    }
	    n++;
	    emat[egTotal][i][j]+=emat[m][i][j];
	    emat[m][i][j]/=nenergy;
	    emat[m][j][i]=emat[m][i][j];
	  }
	}
	emat[egTotal][i][j]/=nenergy;
	emat[egTotal][j][i]=emat[egTotal][i][j];
      }
    }
    if (bFree) {
      if (bRef) {
	fprintf(stderr,"Will read reference energies from inputfile\n");
	neref = get_lines(opt2fn("-eref",NFILE,fnm), &ereflines);
	fprintf(stderr,"Read %d reference energies\n",neref);
	snew(eref, neref);
	snew(erefres, neref);
	for(i=0; (i<neref); i++) {
	  snew(erefres[i], 5);
	  sscanf(ereflines[i],"%s %f",erefres[i],&eref[i]);
	}
      }
      snew(eaver,ngroups);
      for (i=0; (i<ngroups); i++) {
	for (k=0; (k<nenergy); k++)
	  eaver[i] += e[i][k];
	eaver[i] /= nenergy;
      }
      beta = 1.0/(BOLTZ*reftemp);
      snew(efree,ngroups);
      snew(edif,ngroups);
      for (i=0; (i<ngroups); i++) {
	expE=0;
	for (k=0; (k<nenergy); k++) {
	  expE += exp(beta*(e[i][k]-eaver[i]));
	}
	efree[i] = log(expE/nenergy)/beta + eaver[i];
	if (bRef) {
	  n = search_str2(neref,erefres,groups[i]);
	  if (n != -1) {
	    edif[i] = efree[i]-eref[n];
	  } else {
	    edif[i] = efree[i];
	    fprintf(stderr,"WARNING: group %s not found "
		    "in reference energies.\n",groups[i]);
	  }
	} else 
	  edif[i]=0;
      }
    }
   
    emid = 0.0;/*(emin+emax)*0.5;*/
    for(m=0; (m<egNR); m++)
      egrp_nm[m]=egrp_nm[m];
    egrp_nm[egTotal]="total";
    for (m=0; (m<egNR+egSP); m++) 
      if (egrp_use[m]) {
	emin=1e10;
	emax=-1e10;
	for (i=0; (i<ngroups); i++) {
	  for (j=i; (j<ngroups); j++) {
	    if (emat[m][i][j] > emax)
	      emax=emat[m][i][j];
	    else if (emat[m][i][j] < emin)
	      emin=emat[m][i][j];
	  }
	}
	if (emax==emin)
	  fprintf(stderr,"Matrix of %s energy is uniform at %f "
		  "(will not produce output).\n",egrp_nm[m],emax);
	else {
	  fprintf(stderr,"Matrix of %s energy ranges from %f to %f\n",
		  egrp_nm[m],emin,emax);
	  if ((bCutmax) || (emax>cutmax)) emax=cutmax;
	  if ((bCutmin) || (emin<cutmin)) emin=cutmin;
	  if ((emax==cutmax) || (emin==cutmin))
	    fprintf(stderr,"Energy range adjusted: %f to %f\n",emin,emax);
	  
	  sprintf(fn,"%s%s",egrp_nm[m],ftp2fn(efXPM,NFILE,fnm));
	  sprintf(label,"%s Interaction Energies\0",egrp_nm[m]);
	  out=ffopen(fn,"w");
	  if (emin>=emid)
	    write_xpm(out,label,"Energy (kJ/mol)",
		      "Residue Index","Residue Index",
		      ngroups,ngroups,groupnr,groupnr,emat[m],
		      emid,emax,rmid,rhi,&nlevels);
	  else if (emax<=emid)
	    write_xpm(out,label,"Energy (kJ/mol)",
		      "Residue Index","Residue Index",
		      ngroups,ngroups,groupnr,groupnr,emat[m],
		      emin,emid,rlo,rmid,&nlevels);
	  else
	    write_xpm3(out,label,"Energy (kJ/mol)",
		       "Residue Index","Residue Index",
		       ngroups,ngroups,groupnr,groupnr,emat[m],
		       emin,emid,emax,rlo,rmid,rhi,&nlevels);
	  ffclose(out);
	}
      }
    snew(etot,egNR+egSP);
    for (m=0; (m<egNR+egSP); m++) {
      snew(etot[m],ngroups);
      for (i=0; (i<ngroups); i++) {
	for (j=0; (j<ngroups); j++) 
	  etot[m][i]+=emat[m][i][j];
      }
    }

    out=xvgropen(ftp2fn(efXVG,NFILE,fnm),"Mean Energy","Residue","kJ/mol");
    xvgr_legend(out,0,NULL);
    j=0;
    for (m=0; (m<egNR+egSP); m++) 
      if (egrp_use[m])
	fprintf(out,"@ legend string %d \"%s\"\n",j++,egrp_nm[m]);
    if (bFree)
      fprintf(out,"@ legend string %d \"%s\"\n",j++,"Free");
    if (bFree)
      fprintf(out,"@ legend string %d \"%s\"\n",j++,"Diff");
    fprintf(out,"@TYPE nxy\n");
    fprintf(out,"#%3s","grp");
    for (m=0; (m<egNR+egSP); m++)
      if (egrp_use[m])
	fprintf(out," %9s",egrp_nm[m]);
    if (bFree)
      fprintf(out," %9s","Free");
    if (bFree)
      fprintf(out," %9s","Diff");
    fprintf(out,"\n");
    for (i=0; (i<ngroups); i++) {
      fprintf(out,"%3.0f",groupnr[i]);
      for (m=0; (m<egNR+egSP); m++)
	if (egrp_use[m])
	  fprintf(out," %9.5g",etot[m][i]);
      if (bFree)
	fprintf(out," %9.5g",efree[i]);
      if (bRef)
	fprintf(out," %9.5g",edif[i]);
      fprintf(out,"\n");
    }
    fclose(out);
  } else {
    fprintf(stderr,"While typing at your keyboard, suddenly...\n"
	    "...nothing happens.\nWARNING: Not Implemented Yet\n");
/*
    out=ftp2FILE(efMAT,NFILE,fnm,"w");
    n=0;
    emin=emax=0.0;
    for (k=0; (k<nenergy); k++) {
      for (i=0; (i<ngroups); i++)
	for (j=i+1; (j<ngroups); j++)
	  emat[i][j]=eneset[n][k];
      sprintf(label,"t=%.0f ps",time[k]);
      write_matrix(out,ngroups,1,ngroups,groupnr,emat,label,emin,emax,nlevels);
      n++;
    }
    fclose(out);
*/
  }
  if (bEDR)
    xdrclose(&xdr);
  else
    ffclose(in);
  
  thanx(stdout);
  
  return 0;
}
